arrays/arrays.c

```diff
@@ -9,205 +9,152 @@ typedef struct Array {
-  char **p = arr->elements;
-  for(int i=0; i<arr->count; i++)
-  {
-    new_storage[i] = *p++;
-  }
-  // Free the old elements array (but NOT the strings they point to)
-  free(arr->elements);
-  // Update the elements and capacity to new values
-  arr->elements = new_storage;
-  set_rest_null(arr);
-  arr->capacity *= 2;
-  }
-
-
-
-
-
-
+    // Allocate memory for the Array struct
+    Array *arr = malloc(sizeof(Array));
 
+    // Set initial values for capacity and count
+    arr->capacity = capacity;
+    arr->count = 0;
 
+    // Allocate memory for elements
+    arr->elements = calloc(capacity, sizeof(char*));
 
+    return arr;
 
+}
 
+void destroy_array(Array *arr)
+{
+    for (int i = 0; i < arr->count; ++i) {
+        arr->elements[i] = NULL;
+        free(arr->elements[i]);
+    }
 
+    free(arr->elements);
+    free(arr);
+}
 
+void resize_array(Array *arr)
+{
+    char **storage = calloc((2 * arr->capacity), sizeof(char *));
 
+    for (int i = 0; i < arr->count; ++i) {
+        storage[i] = arr->elements[i];
+    }
 
+    for (int j = 0; j < arr->count; ++j) {
+        arr->elements[j] = NULL;
+        free(arr->elements[j]);
+    }
 
+    free(arr->elements);
 
-/************************************
- *
- *   ARRAY FUNCTIONS
- *
- ************************************/
+    arr->elements = storage;
+    arr->capacity = arr->capacity * 2;
+}
 
-//*****
- * Return the element in the array at the given index.
- *
- * Throw an error if the index is out of range.
- *****/
 char *arr_read(Array *arr, int index) {
 
-  // Throw an error if the index is greater than the current count
-  if(index > arr->count)
-  {
-    fprintf(stderr, "invalid index\n");
-    exit(2);
-  }
-  // Otherwise, return the element at the given index
-  else
-  {
-    return arr->elements[index];
-  }
-}
-
-int get_length(char *string)
-{
-  int length = 0;
-  for(int i=0; string[i]!='\0'; i++)
-  {
-    length++;
-  }
-  return ++length;
-}
+    // Throw an error if the index is greater than the current count
+    if(arr->count < index) {
+        printf("Index is higher that current count.");
+        exit(1);
+    }
 
-char* copy_of_string(char* string)
-{
-  int length = get_length(string);
-  char *p = malloc(length);
-  for(int i=0; i<length; i++)
-  {
-    p[i] = string[i];
-  }
-  return p;
+    // Otherwise, return the element at the given index
+    return arr->elements[index];
 }
 
-
-/*****
- * Insert an element to the array at the given index
- *****/
 void arr_insert(Array *arr, char *element, int index) {
 
-  // Throw an error if the index is greater than the current count
+    // Throw an error if the index is greater than the current count
+    if(index > arr->count){
+        printf("Index is higher than current count.");
+        exit(1);
+    }
 
-  // Resize the array if the number of elements is over capacity
+    // Resize the array if the number of elements is over capacity
+    if(arr->count + 1 > arr->capacity) {
```

Comment:
What if the number of elements is at capacity? Should the array be resized before inserting?